### PR TITLE
Rework wasm initialization and reduce global state

### PR DESCRIPTION
**Pull request by cauthmann**
As discussed, I will eventually need two editor instances side-by-side, so we cannot have global state.

This PR is the beginning of that work. Some global state has been reduced or eliminated, but there's still plenty of work to be done.

I will add further commits as I keep working on it, chipping away at the global state one piece at a time. You can merge anytime you like.

My current plan is as follows:
  * Step 1: remove global state from the vue app. `utilities/{dialog/documents,fullscreen}.ts` still contain a lot of them, and I'll need to wrap my head around the response handler logic. On that topic, I'd like to move the stateful code from `utilities/` to `state/` or somewhere, if that's ok with you?
  * Step 2: remove global state from the wasm.

---
This change is [Reviewable](https://reviewable.io/reviews/graphiteeditor/graphite/379)

### CODE REVIEW

**Comment by TrueDoctor** [comment]
Step two is pretty much trivial, let me know if you need help with that

**Comment by cauthmann** [comment]
Step 1 is done. At least the obvious vue state (anything that calls `reactive()`) is no longer global.

Next up is tackling the WASM, but that may have to wait. Feel free to review/merge until here.

**Comment by Keavon** [comment]
Sorry I've taken so long to get to this. Since the change is rather big, I think I will need to wait a few more days until my Thanksgiving break next week. Then I can dive in over a couple days and give this review the attention it deserves. I very much do not intend to make such delayed reviews a regular thing and I offer my sincerest apologies that this has had to slip my attention lately. School has been a large focus out of necessity, but the end is in sight. Very soon— full steam ahead on Graphite!

File: frontend/src/utilities/documents.ts
Comment (mfish33) [reply frontend/src/utilities/documents.ts:43]: Nice catch this change came in after tub created the pr so it was missing from the document state file

File: frontend/src/utilities/documents.ts
Comment (mfish33) [reply frontend/src/utilities/documents.ts:127]: This is currently been migrated to document state

File: frontend/src/state/js-dispatcher.ts
Comment (mfish33) [reply frontend/src/state/js-dispatcher.ts]: This is not an unsafe cast. It is a cast from a specific union of exact types to a more generic type. It prevents typescript from complaining since it does not want to do the coercion automatically.

File: frontend/src/App.vue
Comment (mfish33) [reply frontend/src/App.vue:246]: removed

File: frontend/src/state/js-dispatcher.ts
Comment (cauthmann) [reply frontend/src/state/js-dispatcher.ts]: `let a = b as X` is a type assertion. Like `any`, you're asking the compiler to trust you. It's unsafe.

`let a: X = b` is safe and checked.

https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/docs/rules/consistent-type-assertions.md has an example (and is probably a good rule to add with all options set to `'never'`):

> `const x = {} as { foo: number }` will succeed

**Comment by mfish33** [comment]
I think this is almost ready to go. There is still a class inside `documents.ts` `DocumentSaveState` which will be moved to js-messages in my next PR

File: frontend/src/App.vue
Comment (Keavon) [inline frontend/src/App.vue:241]: Why is this optional? Please comment to avoid confusion (or fix it if it's better being not an optional parameter).

File: frontend/src/state/js-dispatcher.ts
Comment (Keavon) [reply frontend/src/state/js-dispatcher.ts]: @mfish33 I'm still not understanding your description of this as "not an unsafe cast". To the extent I understand my typescript, this is an unsafe cast. (See cauthmann's description above also.) If the TypeScript compiler can't figure out the types on its own, that means something is wrong with the types here. Do we need to specify the types of `export const messageConstructors = {` in `js-messages.ts`?

File: frontend/src/App.vue
Comment (cauthmann) [reply frontend/src/App.vue:241]: Because the inputManager is not initialized in `data()` (because the div hasn't been created yet), but in `mounted()`. So there is a time in the lifecycle of the component where `inputManager = undefined`, and the type must reflect that.

Comment (Keavon) [reply frontend/src/App.vue:241]: Gotcha, I'll add that comment now then, thanks for explaining.

**Comment by cauthmann** [comment]
I've pushed a commit that should fix the unsafe cast, but due to the state of my dev environment, the only tests I've performed is looking for errors in my IDE. Please do a quick smoke test before merging.

**Comment by cauthmann** [comment]
As to the rest of the commits, I've read everything and didn't find anything problematic. I really like 5bba08b! From my point of view, this looks ready to merge.

Thanks to both of you for taking the time to work on this!

**Comment by Keavon** [comment]
> I've pushed a commit that should fix the unsafe cast, but due to the state of my dev environment, the only tests I've performed is looking for errors in my IDE. Please do a quick smoke test before merging.

This looks like a nice commit, thanks! Unfortunately, though, mouse events aren't working so I can't drag to create shapes, pan to zoom around the canvas, or <kbd>Ctrl</kbd>+mousewheel zoom (this isn't `preventDefault`ed and it lets the browser adjust the web page scale).

I had a quick look and didn't spot an obvious reason for the bug, but perhaps when @mfish33 wakes up he can take a look and debug it. I'm too tired at this point to keep debugging.

I also just finished my last pass through with another full round of code reviews and made a few comments, plus small tweaks to the code which I'm going to commit now.

@cauthmann if you'd like, I can give you write access to the repo and you can push the code to a temporary testing branch and let Cloudflare Pages give you a build, although the downside is that it takes about 15 minutes to get a build.

**Comment by cauthmann** [comment]
Oh dear. Please check if this commit fixes it.

**Comment by Keavon** [comment]
That works, thanks for the fix. Except one issue still isn't fixed:

> Ctrl+mousewheel zoom (this isn't `preventDefault`ed and it lets the browser adjust the web page scale).

File: frontend/src/utilities/input.ts
Comment (Keavon) [reply frontend/src/utilities/input.ts:116]: This explains the issue above. I'll reinstate this.

Comment (Keavon) [reply frontend/src/utilities/input.ts:116]: I changed this. However it didn't seem to fix the fact that I can Ctrl+mousewheel to adjust page zoom. I can look at when I wake up if one of you don't figure it out first. (@cauthmann you can add this repo as a remote origin and push to a branch of your choosing, then you should get a build link from Cloudflare Pages within 15 minutes if you want to test a build.)

File: frontend/src/dispatcher/js-messages.ts
Comment (mfish33) [inline frontend/src/dispatcher/js-messages.ts:144]: I do not think it is worth doing runtime checks as we should be able to assume that the rust side will send the correct data

File: frontend/src/state/documents.ts
Comment (mfish33) [reply frontend/src/state/documents.ts:37]: You are correct must have missed it when merging

File: frontend/src/state/wasm-loader.ts
Comment (mfish33) [reply frontend/src/state/wasm-loader.ts:56]: I split up the code into two lines and I think it is just easier to read. The comment is no longer necessary.

File: frontend/src/utilities/input.ts
Comment (mfish33) [reply frontend/src/utilities/input.ts:116]: I messed up when changing the code to this structure and added passive: true instead of false causing the preventDefault from working

File: frontend/src/dispatcher/js-messages.ts
Comment (Keavon) [reply frontend/src/dispatcher/js-messages.ts:144]: Agreed, unless it's likely that a bug can arise during maintenance in the future and this helps explain what's wrong in debugging. If that doesn't seem likely, go ahead and remove it.

Comment (Keavon) [reply frontend/src/dispatcher/js-messages.ts:144]: This is also necessary to satisfy TypeScript. I'm going to leave it for now but feel free to improve it in a later PR.

**Comment by Keavon** [comment]
Here are some bugs that will need to be resolved that I found when QA testing this now:

* Layers don't show up in the layer panel at all
* After a layer gets selected by clicking on it (but not with box selection), keyboard shortcuts like <kbd>Delete</kbd> and <kbd>Shift</kbd><kbd>X</kbd> (swap primary/secondary colors) and <kbd>G</kbd>/<kbd>R</kbd>/<kbd>S</kbd> (grab/rotate/scale) and hotkeys for switching tools and <kbd>Ctrl</kbd><kbd>1</kbd>/<kbd>Ctrl</kbd><kbd>2</kbd> (zoom to 100%/200% which should preventDefault so the browser doesn't switch web page tabs but that doesn't work right now) don't work, until clicking outside the viewport area in some of the HTML elements outside it or <kbd>Shift</kbd>-clicking somewhere else in the canvas.  This also seems to affect other tools, not just the Select Tool.
* After causing a panic (*Help > Debug: Panic (DANGER)*), every subsequent mouse movement spams it to the console again.
* The panic proxy [isn't always working](https://files.keavon.com/-/FuzzyHastyCornsnake/capture.png). To reproduce: create a shape, save it to a `.graphite` file, open that file, then use the Select Tool to click on the shape. This panics in master also, but in this branch it shows the `Uncaught RuntimeError: unreachable` which the panic proxy is designed to suppress. (Because "reachable" is a concept in both Rust and wasm and it's confusing to have two messages show up where many developers will likely assume this means unreachable in Rust, but it's actually a wasm assembly instruction called `unreachable`.)
* *File > Close* fails and puts this in the console: `While handling FrontendMessage ""DisplayConfirmationToCloseDocument"", JavaScript threw an error: JsValue(Error: recursive use of an object detected which would lead to unsafe aliasing in rust`

**Comment by Keavon** [comment]
@cauthmann you'll want to take a look at ae18ce8ee340c06c1913efc0b088f2d86d365f10 in which I reverted a couple changes you made, probably at the expense of your multi-editor goal. When you get the chance to run the code on your machine, please submit another (thoroughly tested) PR to address this. But we need to merge this PR now with the regression fixes in that commit linked above.

**Review by Keavon — APPROVED** [review]

**Comment by cauthmann** [comment]
Re: The runtime checks in the message code.

Unless we automatically generate message types from rust code (using ts-rs or similar), then any changes to the rust side will cause breakage on the javascript side, and it's better to have a cheap runtime check with a proper error message than some semi-random error with an unhelpful description somewhere later.

Especially with code that accesses and parses random wasm memory, I'd prefer to fail before that memory is touched.

This isn't meant to protect against the rust side turning evil, it's meant to protect against commits like 4494993 (and others in this PR), which break stuff in ways that remained undetected.

**Comment by cauthmann** [comment]
> @cauthmann you'll want to take a look at [ae18ce8](https://github.com/GraphiteEditor/Graphite/commit/ae18ce8ee340c06c1913efc0b088f2d86d365f10) in which I reverted a couple changes you made, probably at the expense of your multi-editor goal.

Putting all the event handlers back onto the window will break that entirely, yes.

**Comment by Keavon** [comment]
> Re: The runtime checks in the message code.
> 
> Unless we automatically generate message types from rust code (using ts-rs or similar), then any changes to the rust side will cause breakage on the javascript side, and it's better to have a cheap runtime check with a proper error message than some semi-random error with an unhelpful description somewhere later.
> 
> Especially with code that accesses and parses random wasm memory, I'd prefer to fail before that memory is touched.
> 
> This isn't meant to protect against the rust side turning evil, it's meant to protect against commits like [4494993](https://github.com/GraphiteEditor/Graphite/commit/44949937c9cd80e75fe76032cd6d6c69dd0609f1) (and others in this PR), which break stuff in ways that remained undetected.

@mfish33 thoughts?

**Pull request merged**
